rust/datafusion/src/execution/context.rs

```diff
@@ -375,15 +380,39 @@ impl ScalarFunctionRegistry for ExecutionContext {
     }
 }
 
+/// Provides OptimizerRule instances to
```

Comment:
I am not a huge fan of this `OptimizerRuleSource` but it was the best I could come up with at this point

Reply:
I think I would prefer `Provider` to `Source` in this context. Also, would be good to document whether these are logical or physical optimizer rules.

Reply:
As it turns out your suggestion for a single planner trait has removed the need for this trait entirely 🎉